### DISCUSSION TITLE
Get values of RSSI - signal strength between ESPnow hub and ESPnow end device

### DISCUSSION THREAD

**Discussion by homonto** [HomeSpan/HomeSpan#513] [General]
Hi,

is there any way that when end device sends the data to the Hub (ESPnow), the hub exposes the rssi of that connection (not the rssi between the hub and the wifi router)?

I implemented such functionality in my ESPnow implementation but I would love to see if the same can be done here, in the HomeKit world.

I have done this this way:
`esp_wifi_set_promiscuous_rx_cb(&promiscuous_rx_cb);`
in the start wifi() function and then inside it:
```
void promiscuous_rx_cb(void *buf, wifi_promiscuous_pkt_type_t type)
{
  if (type != WIFI_PKT_MGMT)
    return;
  const wifi_promiscuous_pkt_t *ppkt = (wifi_promiscuous_pkt_t *)buf;
  myData_aux.rssi = ppkt->rx_ctrl.rssi;
}
```

so rssi is inside my myData_aux.rssi struct and I can see how good/bad signal strength is between these 2 devices.

Of course I am talking about having this functionality WITHOUT MODIFYING your library manually.

Thank you

**Comment by homonto** [comment]
OK, this point (RSSI from remote devices) seems not important for you, but for me it is - I have many sensors around the house and sometimes I am missing data from some of them - now I know why - poor RSSI.
I have added a bit of dirty code to your library and got it working - unfortunately I am not good at C++ to have it done proper way so if you have nothing else to do and you find it attractive maybe you can add it to the library.

1. So first: **promiscuous mode has to be enabled** with the call back function (as I described above). 
2. Second: we need a variable that can store the measured RSSI.

Since I don't know where to put the variable that can be common for both promiscuous_rx_cb() and void SpanPoint::dataReceived() I declared it at the beginning of HomeSpan.cpp

`int zh_rssi;`

as well as included the header file with the data structure being sent by remote device is (I will explain why it is needed below):

`#include "data_structure.h"`

then I defined the proper function (declaration I put in HomeSpan.h)

```
void promiscuous_rx_cb(void *buf, wifi_promiscuous_pkt_type_t type)
{
  if (type != WIFI_PKT_MGMT)
    return;
  const wifi_promiscuous_pkt_t *ppkt = (wifi_promiscuous_pkt_t *)buf;
  zh_rssi = (int)ppkt->rx_ctrl.rssi;
}
```

Then inside SpanPoint::SpanPoint() (both, as you have one overload I think) but BEFORE WiFi.mode():

```
  esp_wifi_set_promiscuous(0);
  esp_wifi_set_promiscuous_rx_cb(&promiscuous_rx_cb);
  esp_wifi_set_promiscuous(1);
```

Finally, in void SpanPoint::dataReceived() after all ifs:

```
  struct_message * ptr = (struct struct_message *) incomingData;
  ptr->md_rssi = zh_rssi;
```

where:

- **struct_message** is the structure of my data being sent
- **zh_rssi** is the global variable that received the value inside the call back function (promiscuous_rx_cb)
- and finally, **ptr** is the pointer to the data received, but it is casted to proper structure.

This way, knowing in which field in the data structure, there is a place where I can have rssi, I replace this value (it is empty btw from the remote sensor, as remote sensor has no idea about it) with the value received during the transmission between the bridge and the remote.

This way, without modifying the **incomingData**, I was able to insert there measured RSSI of the signal from remote to the bridge.
The next line remains yours:

`xQueueSend((*it)->receiveQueue, incomingData, 0); `


Now, with regards to the data structure, in my case it looks like this:

```
typedef struct struct_message         
{
  float md_temp;
  float md_hum;
  float md_light;
  uint8_t md_bat;
  char md_version[20];
  uint8_t md_mcu_model;
  uint8_t md_charging;
  uint8_t md_boot_reason;
  uint8_t md_wakeup_reason;
  unsigned long md_bootcount;

  // added for future use:
  uint8_t var1;
  uint8_t var2;
  int8_t md_rssi;
  int8_t var4;
  int8_t var5;
  char var6[10];
  char var7[22];
  char var8[32];
  float md_bat_volts;
  float var10;
  float var11;
  float var12;
  float var13;
  int32_t var14;
  int32_t var15;
  uint32_t var16;
  uint32_t var17;

} struct_message;
```

About the structure: your code disregards data received, if the arrived amount of bytes is different from expected one - that is perfectly fine, provided, you DON'T want to change the structure ever, or, you have easy access to the remote sensors to reprogram them. 
That was not a case for me so I modified one more place in your library - just before this line:

`struct_message * ptr = (struct struct_message *) incomingData;`

I added these ones:

```
    #if (ALLOW_DATA_SIZE_MISMATCH == 1)
    {
      LOG1("RECEIVED DATA SIZE MISMATCH BUT WILL BE USED!\n");
    } else 
    {
      LOG1("RECEIVED DATA SIZE MISMATCH AND WILL BE IGNORED!\n");
      return;
    }
    #endif
```
and of course 

`#define ALLOW_DATA_SIZE_MISMATCH 1`

in my config file (for the time being, until all remote sensors are with updated firmware with new structure. Then I will change to 0).

Outcome?

<img width="788" alt="Screenshot 2023-12-28 at 16 24 14" src="https://github.com/HomeSpan/HomeSpan/assets/46562447/f0290f37-c4f6-40b1-9679-02ad991d9ca9">

Is it worth your time? I don't know ;-)
But maybe someone will use it. 
The only problem today is: my modifications in your code the way I did it, made me disconnected from the original library.

**Reply by anlpereira** [reply]
Hi Homonto, can you share the full code? Thanks

**Reply by homonto** [reply]
hi @anlpereira which part you are interested in: RSSI?

**Reply by anlpereira** [reply]
Thanks, is this graphic chart made for Homekit?

**Reply by SokDuong** [reply]
hi @homonto can u share your full code ?
thank

**Reply by SokDuong** [reply]
esp_wifi_set_promiscuous(0);
  esp_wifi_set_promiscuous_rx_cb(&promiscuous_rx_cb);
  esp_wifi_set_promiscuous(1);
   
Do we need to include other library then esp_now.h to run that function ?

**Reply by homonto** [reply]
no

**Reply by homonto** [reply]
I need to first sanitize because I put the passwords file into the project so gimme a time and I will

**Reply by anlpereira** [reply]
Thank you very much.

**Reply by ve7kilohertz** [reply]
HI,

Excellent work on getting RSSI to work with ESP-NOW. Just wondering if you finished cleaning up the code and could you please share it? I have a remote water tank that I am having difficulty staying connected to and a variety of antennas to try, having RSSI would be a HUGE help.

Cheers

Paul

**Reply by homonto** [reply]
> HI,
> 
> Excellent work on getting RSSI to work with ESP-NOW. Just wondering if you finished cleaning up the code and could you please share it? I have a remote water tank that I am having difficulty staying connected to and a variety of antennas to try, having RSSI would be a HUGE help.
> 
> Cheers
> 
> Paul

try this link please:
https://gitea.papio.uk/zyghom/public-Homekit_Split

let me know if I shared anything secret or so ;-)

**Comment by ve7kilohertz** [comment]
Thank you so much for the quick reply. I am just fighting with the AJ-SR04M sensor right now, trying to get the readings stabilized. It’s in an 8’ diameter water tank about 5’ from the surface.

 

I’ll look at your code shortly.

 

Thank you again!



Paul

**Comment by ve7kilohertz** [comment]
Hello Homonto, I am back on the project and had a look at the link you sent me, thank you. I may have asked for the wrong thing, I was actually looking for the ESP-NOW code you had working for your non-Home-Span project, something that would work on a NODEMCU type ESP8266 or similar. I just need to have RSSI between my remote water tank sender (Wemos D1 mini PRO) and my receiver at the house NODEMCU. Thank you again. Paul

**Comment by homonto** [comment]
@ve7kilohertz  here you are:
https://gitea.papio.uk/zyghom/public-espnow2mqtt
